proposals/0239-operator-designated-types.md

```diff
@@ -0,0 +1,137 @@
+- `-Xfrontend -solver-enable-operator-designated-types`: enables having the constraint solver make use of the declarations
+
+The current implementation has been tested across 25 test cases that were known to be slow and/or result in the type checker giving up after expending substantial effort attempting to type check the expression.
+
+In addition to enabling this new *designated type* functionality, this testing disables the `shrink` portion of the expression type checker (via `-Xfrontend -solver-disable-shrink`).  This portion of the expression type checker is known to help occassionally but often results in substantially increasing the expression type checking time since it frequently doesn't provide any benefit but does a lot of work. We also disable a number of other problematic hacks (via `-Xfrontend -disable-constraint-solver-performance-hacks`).
+
+In 14 cases, the expressions in question now typecheck in less than 30ms (often as little as 5ms). In the 11 remaining cases, the expression is either one that would fail to typecheck if we were able to typecheck it quickly, or involves other things that are known to be problematic, like closures. 
+
+
+Some illustrative examples:
+
+```
+// Was "too complex", now 13ms:
+let i: Int? = 1
+let j: Int?
+let k: Int? = 2
+
+let _ = [i, j, k].reduce(0 as Int?) {
+  $0 != nil && $1 != nil ? $0! + $1! : ($0 != nil ? $0! : ($1 != nil ? $1! : nil))
+}
+
+
+// Was "too complex", now 5ms:
+func test(a: [String], b: String, c: String) -> [String] {
+  return a.map { $0 + ": " + b + "(" + c + $0 + ")" }
+}
+
+// Was "too complex", now 25ms:
+func test(strings: [String]) {
+  for string in strings {
+    let _ = string.split(omittingEmptySubsequences: false) { $0 == "C" || $0 == "D" || $0 == "H" || $0 == "S"}
+  }
+}
+
+// Was "too complex", now 5ms:
+func test(n: Int) -> Int {
+  return n == 0 ? 0 : (0..<n).reduce(0) {
+    ($0 > 0 && $1 % 2 == 0) ? ((($0 + $1) - ($0 + $1)) / ($1 - $0)) + (($0 + $1) / ($1 - $0)) : $0
+  }
+}
+
+// Was "too complex", now 2ms:
+let a = 1
+_ = -a + -a - -a + -a - -a
+
+```
+
+A number (25+) of other test cases were previously compiling quickly were also tested using this functionality and with the existing hacks disabled, and those cases are as fast or faster than previously.
+
+## Source compatibility
+
+Because of the fact that we fall back to behavior that is very similar to the current behavior of the type checker (exhaustively testing combinations of operators in the cases), this change is very compatible with the current behavior. 
+
+One failure was seen in the source compatibility suite when running with this new behavior enabled by default in all language modes. The failure was in `BlueSocket`, and that same failure also manifests if you build the affected code with `-swift-version 4`. Specificially, for `let mask = 1 << (2 as Int32)`, we now infer `mask` as `Int` rather than `Int32`.
```

Comment:
Specificially => Specifically

```suggestion
One failure was seen in the source compatibility suite when running with this new behavior enabled by default in all language modes. The failure was in `BlueSocket`, and that same failure also manifests if you build the affected code with `-swift-version 4`. Specifically, for `let mask = 1 << (2 as Int32)`, we now infer `mask` as `Int` rather than `Int32`.
```